### PR TITLE
Added chpldocs+support for distributions, tuples

### PR DESCRIPTION
**Pull request by vasslitvinov**
This PR replaces #2583 
for clean revision history

For distributions, given it is code freeze so we do not get to modify
'chpldoc', I took the approach - suggested by Elliot - to put all
documentation into the comment associated with the distribution class decl.

fixDistDocs.perl ensures that no other decls for each file
make it into the html page.

I ended up with a style of presenting each distribution that's somewhere
between a formal spec and a narrative-like tutorial. We can revisit that.

The diffs for ReplicatedDist.chpl show a couple of param decls moved.
That's just because of 'diff' - there are no code changes,
at least not intentional.

I listed some chpldoc TODOs in each file. Most notably:

* Move the notes about constructors and selected other functions
from that monolithic comment to the comment on the actual declaration.

  There will be a challenge because some constructors use the functions
like _determineRankFromStartIdx() and _determineIdxTypeFromStartIdx()
that we might not want the users to know about.

  The benefit of such a move is that the documentation will be kept
in sync with the code.

* I rephrased and toned down this piece from the spec - OK?

   It is parameterized by the rank and index type of the domains it
   supports.  Thus domains of different ranks or different index types
   must be distributed with different Cyclic [or whatever] distributions.

* Need automatic testing of examples that appear in chpldocs.

Separately, we have been considering chpldocs for Tuples to replace
the section "Predefined Functions and Methods on Tuples" in the spec.
For now we decided to keep that section and at the same time
provide chpldocs for those functions.

### CODE REVIEW

File: modules/Makefile [hunk 1]
```diff
@@ -86,7 +86,20 @@ MODULES_TO_DOCUMENT = \
 	standard/UtilReplicatedVar.chpl \
 	$(SYS_CTYPES_MODULE_DOC)
 
+DISTS_TO_DOCUMENT = \
```
Comment (ronawho) [inline modules/Makefile]: Do you mind alphabetizing this and INTERNAL_MODULES_TO_DOCUMENT

File: modules/Makefile [hunk 2]
```diff
@@ -94,10 +107,14 @@ INTERNAL_MODULES_TO_DOCUMENT = \
...
 	$(CHPLDOC) --save-sphinx ${MODULE_SPHINX} $(MODULES_TO_DOCUMENT) \
-	$(INTERNAL_MODULES_TO_DOCUMENT)
+	$(DISTS_TO_DOCUMENT) $(INTERNAL_MODULES_TO_DOCUMENT)
 	./internal/fixInternalDocs.sh ${MODULE_SPHINX}
+	./dists/fixDistDocs.perl      ${MODULE_SPHINX}
 	cp -rf ${MODULE_SPHINX}/source/modules/standard ${DOC_SPHINX}/source/modules/
+	cp -rf ${MODULE_SPHINX}/source/modules/dists    ${DOC_SPHINX}/source/modules/
+	cp -rf ${MODULE_SPHINX}/source/modules/layouts  ${DOC_SPHINX}/source/modules/
 	cp -rf ${MODULE_SPHINX}/source/modules/internal ${DOC_SPHINX}/source/modules/
+	rm -f  ${DOC_SPHINX}/source/modules/dists/DSIUtil.rst
```
Comment (ronawho) [inline modules/Makefile]: Can this be done inside fixDistDocs?

File: modules/dists/fixDistDocs.perl [hunk 3]
```diff
@@ -0,0 +1,80 @@
+#!/usr/bin/env perl
```
Comment (ronawho) [inline modules/dists/fixDistDocs.perl]: I didn't review this in depth, but it looks ok to me.

File: modules/Makefile (hunk 1 above)
Comment (ronawho) [inline modules/Makefile]: Actually, can you just alphabetize DIST_TO_DOCUMENT, I'll take care of INTERNAL_MODULES_TO_DOCUMENT in an upcoming PR

File: modules/internal/ChapelTuple.chpl [hunk 4]
```diff
@@ -76,18 +92,29 @@ module ChapelTuple {
...
+  
+  pragma "no doc"
   proc isHomogeneousTupleValue(x: _tuple) param
     return __primitive("is star tuple type", x);
   
+  /*
+    Returns `true` if its argument is a tuple type.
+    The argument must be a type.
+  */
   proc isTupleType(type t) param
```
Comment (ronawho) [inline modules/internal/ChapelTuple.chpl]: I think we've been no-doc'ing the is*Type functions too, and using isTuple, isHomogeneousTuple as the public facing functions.

File: modules/internal/ChapelTuple.chpl [hunk 5]
```diff
@@ -114,10 +142,20 @@ module ChapelTuple {
     return __primitive("get svec member", this, i);
   }
 
+  /*
```
Comment (ronawho) [inline modules/internal/ChapelTuple.chpl]: I don't think this should be publicly documented. If anything I think you should add a module level comment about the `--[no-]warn-tuple-iteration` flag with this code example. That flag ultimately controls this param, and so I don't think we want users to be aware of it

File: modules/internal/ChapelTuple.chpl (hunk 4 above)
Comment (ronawho) [inline modules/internal/ChapelTuple.chpl]: Hmm, although the are in the spec.. Maybe check with Brad?

File: modules/internal/ChapelTuple.chpl [hunk 6]
```diff
@@ -17,10 +17,26 @@
...
+
+Besides the functions defined here, the Chapel Language specification
+defines other operations available on tuples: indexing, iteration,
+assignment, and unary, binary, and relational operators.
+
+The following method is also availble:
+
+  .. code-block:: chapel
+
+    proc tuple.size param
```
Comment (ronawho) [inline modules/internal/ChapelTuple.chpl]: Out of curiosity, do you know where the implementation of size lives? Is it in some other module, or in the compiler?

**Comment by ronawho** [comment]
Tuple and infrastructure stuff looks pretty good to me, I left a few comments inline

**Comment by vasslitvinov** [comment]
I followed up on the above feedback and merged #2634 
instead of this, for clean revision history.
Closing this.

**Pull request closed without merging**